docs: internal connection pool for Spring apps (#869)
awslabs/aws-advanced-jdbc-wrapper: docs/using-the-jdbc-driver/using-plugins/UsingTheReadWriteSplittingPlugin.md

When a Statement or ResultSet is created, it is internally bound to the database connection established at that moment. There is no standard JDBC functionality to change the internal connection used by Statement or ResultSet objects. Consequently, even if the read/write plugin switches the internal connection, any Statements/ResultSets created before this will continue using the old database connection. This bypasses the desired functionality provided by the plugin. To prevent these scenarios, an exception will be thrown if your code uses any Statements/ResultSets created before a change in internal connection. To solve this problem, please ensure you create new Statement/ResultSet objects after switching between the writer/reader.

### Session state

The plugin supports session state transfer when switching connection. All attributes mentioned in [Session State](../SessionState.md) are automatically transferred to a new connection. 

### Limitations when using Spring Boot/Framework

#### @Transactional(readOnly = True)

> [!WARNING]\

> > The use of read/write splitting with the annotation @Transactional(readOnly = True) is **not** recommended for any configurations except a configuration with internal connection pool. Other configurations will cause a significant performance degradation.
> 

Reviewer: ```suggestion
> The use of read/write splitting with the annotation @Transactional(readOnly = True) is **only** recommended for configurations using an internal connection pool. Using the annotation with any other configurations will cause a significant performance degradation.
```

When a method with this annotation is hit, Spring calls conn.setReadOnly(true), executes the method, and then calls setReadOnly(false) to restore the connection's initial readOnly value. Consequently, every time the method is called, the plugin switches to the reader, executes the method, and then switches back to the writer. Although the reader connection will be cached after the first setReadOnly call, there is still some overhead when switching between the cached writer/reader connections. This constant switching is not an ideal use of the plugin because it is frequently incurring this overhead. The suggested approach for this scenario is to avoid loading the read/write splitting plugin and instead use the writer cluster URL for your write operations and the reader cluster URL for your read operations. By doing this you avoid the overhead of constantly switching between connections while still spreading load across the database instances in your cluster. 

#### Internal connection pools

> We recommend that you do not enable internal connection pools when using Spring with default datasource configuration. This is because Spring by default uses its own external connection pool. The use of both internal and external pools is not tested and may result in problematic behavior. A Spring application should enable internal connection pool only. That's a recommended configuration.

Reviewer: ```suggestion
We recommend that you **disable** internal connection pools when using Spring with default datasource configuration. Spring by default uses its own external connection pool. The use of both internal and external pools has not been tested and may result in problematic behaviour. A Spring application should enable internal connection pool only. That's a recommended configuration.
```

Reviewer: I don't understand the last sentence. `A Spring application should enable internal connection pool only. That's a recommended configuration.` but the first sentence said to not use internal connection pool when using Spring with default datasource configuration.

Author: I rewrote this part. I hope it's a bit cleaner now.

> We recommend that you do not enable internal connection pools when using Spring with default datasource configuration. This is because Spring by default uses its own external connection pool. The use of both internal and external pools is not tested and may result in problematic behavior. A Spring application should enable internal connection pool only. That's a recommended configuration.

Author: I rewrote this part. I hope it's a bit cleaner now.

> We recommend that you explicitly disable internal connection pools when using Spring. You can check `spring.datasource.type` property to ensure that an external connection pooling is disabled. By default, Spring uses datasource auto-detection, and it may enable external connection pool. The use of both internal and external pools has not been tested and may result in problematic behaviour. A Spring application should enable internal connection pool only. That's a recommended configuration.

Reviewer 2: Does this paragraph mean that internal connection pooling isn't recommended to use, but if they are used, users should disable Spring's external connection pooling? And that both internal and external connection pooling shouldn't be used together? If that's the case, how does this sound:
```suggestion
We recommend that you explicitly disable internal connection pools when using Spring. However, if you do use the driver's internal connection pooling, you can check the `spring.datasource.type` property to ensure that any external connection pooling is disabled. This is necessary because Spring uses datasource auto-detection by default, and it may enable external connection pooling. Using internal and external pools at the same time has not been tested and may result in problematic behaviour. The recommended configuration for a Spring application should either enable internal connection pooling or external connection pooling, but not both at once.
```

> We recommend that you explicitly disable internal connection pools when using Spring. You can check `spring.datasource.type` property to ensure that an external connection pooling is disabled. By default, Spring uses datasource auto-detection, and it may enable external connection pool. The use of both internal and external pools has not been tested and may result in problematic behaviour. A Spring application should enable internal connection pool only. That's a recommended configuration.

Author: ```suggestion
If you want to use the driver's internal connection pooling, we recommend that you explicitly disable external connection pools (provided by Spring). You need to check the `spring.datasource.type` property to ensure that any external connection pooling is disabled. This is necessary because Spring uses datasource auto-detection by default, and it may enable external connection pooling. Using internal and external pools at the same time has not been tested and may result in problematic behaviour. The recommended configuration for a Spring application should either enable internal connection pooling or external connection pooling, but not both at once.
```

> Spring application is encouraged to use configuration profiles and presets optimized for Spring applications. More details are available at [Configuration Presets](../ConfigurationPresets.md).

Reviewer 2: ```suggestion
Spring applications are encouraged to use configuration profiles and presets optimized for Spring applications. More details are available at [Configuration Presets](../ConfigurationPresets.md).
```

> Spring applications are encouraged to use configuration profiles and presets optimized for Spring applications. More details are available at [Configuration Presets](../ConfigurationPresets.md).

Reviewer 2: ```suggestion
Spring applications are encouraged to use configuration profiles and presets optimized specifically for Spring applications. More details are available at [Configuration Presets](../ConfigurationPresets.md).
```

## Example
[ReadWriteSplittingPostgresExample.java](../../../examples/AWSDriverExample/src/main/java/software/amazon/ReadWriteSplittingPostgresExample.java) demonstrates how to enable and configure read/write splitting with the Aws Advanced JDBC Driver.

> [SpringHibernateBalancedReaderOneDataSourceExample](../../../examples/SpringHibernateBalancedReaderOneDataSourceExample/README.md) demonstrates how to enable and configure read/write splitting with the Aws Advanced JDBC Driver. This example application uses a configuration with internal connection pool and provides a load-balanced reader connection according to a specified reader selection strategy. `Read/Write Splitting Plugin` and `@Transactional(readOnly = True)` play in collaboration in this example.

Reviewer: ```suggestion
[SpringHibernateBalancedReaderOneDataSourceExample](../../../examples/SpringHibernateBalancedReaderOneDataSourceExample/README.md) demonstrates how to enable and configure Read/Write Splitting plugin with the AWS JDBC Driver. This example application uses a configuration with internal connection pool and provides a load-balanced reader connection according to a specified reader selection strategy. `Read/Write Splitting Plugin` and `@Transactional(readOnly = True)` play in collaboration in this example.
```

Reviewer 2: Adding on top of Karen's suggestion:
```suggestion
[SpringHibernateBalancedReaderOneDataSourceExample](../../../examples/SpringHibernateBalancedReaderOneDataSourceExample/README.md) demonstrates how to enable and configure the Read/Write Splitting plugin with the AWS JDBC Driver. This example application uses a configuration with internal connection pooling and provides a load-balanced reader connection according to a specified reader selection strategy. The `Read/Write Splitting Plugin` and `@Transactional(readOnly = True)` annotation are both used in this example.
```
I'm not entirely sure what the last sentence means, does this sound right?

Author: I've changed this sentence. Please take a look.

> [SpringHibernateBalancedReaderOneDataSourceExample](../../../examples/SpringHibernateBalancedReaderOneDataSourceExample/README.md) demonstrates how to enable and configure Read/Write Splitting plugin with the AWS JDBC Driver. This example application uses a configuration with internal connection pool and provides a load-balanced reader connection according to a specified reader selection strategy. `@Transactional(readOnly = True)` annotations in the code give a hint for `Read/Write Splitting Plugin` for switching between datasources.

Reviewer 2: ```suggestion
[SpringHibernateBalancedReaderOneDataSourceExample](../../../examples/SpringHibernateBalancedReaderOneDataSourceExample/README.md) demonstrates how to enable and configure Read/Write Splitting plugin with the AWS JDBC Driver. This example application uses a configuration with internal connection pooling and provides a load-balanced reader connection according to a specified reader selection strategy. `@Transactional(readOnly = True)` annotations in the code help the `Read/Write Splitting Plugin` switch between datasources.
```

> [SpringHibernateBalancedReaderTwoDataSourceExample](../../../examples/SpringHibernateBalancedReaderTwoDataSourceExample/README.md) demonstrates how to enable and configure read/write splitting with the Aws Advanced JDBC Driver. This example application uses a configuration with two Spring datasources where each of them uses internal connection pool. The example application provides a load-balanced reader connection according to a specified reader selection strategy. The example application doesn't use `Read/Write Splitting Plugin`. Switching between writer datasource and reader datasource occurs with the help of `@WithLoadBalancedReaderDataSource` annotation. 

Reviewer: ```suggestion
[SpringHibernateBalancedReaderTwoDataSourceExample](../../../examples/SpringHibernateBalancedReaderTwoDataSourceExample/README.md) demonstrates how to enable and configure Read/Write Splitting plugin with the AWS JDBC Driver. This example application uses a configuration with two Spring datasources where each of them uses internal connection pool. The example application provides a load-balanced reader connection according to a specified reader selection strategy. The example application doesn't use `Read/Write Splitting Plugin`. Switching between writer datasource and reader datasource occurs with the help of `@WithLoadBalancedReaderDataSource` annotation. 
```

Reviewer 2: Adding onto Karen's suggestion:
```suggestion
[SpringHibernateBalancedReaderTwoDataSourceExample](../../../examples/SpringHibernateBalancedReaderTwoDataSourceExample/README.md) demonstrates how to enable and configure Read/Write Splitting plugin with the AWS JDBC Driver. This example application uses a configuration with two Spring datasources, where each datasource uses internal connection pooling. The example application provides a load-balanced reader connection according to a specified reader selection strategy. The example application does not use the `Read/Write Splitting Plugin`. Switching between the writer datasource and reader datasource occurs with the help of the `@WithLoadBalancedReaderDataSource` annotation.
```
